packages/editor/src/components/provider/use-block-editor-settings.js

```diff
@@ -268,6 +273,7 @@ function useBlockEditorSettings( settings, postType, postId, renderingMode ) {
 				)
 			),
 			[ globalStylesDataKey ]: globalStylesData,
+			[ globalStylesLinksDataKey ]: globalStylesLinksData,
```

Comment:
I need to catch up with discussion here, but I'm wary of adding another global styles property. 

I suppose it's safer than blindly passing on everything from `mergedGlobalStyles`, but wondering if they should be combined into a `{ styles, settings, _links }` object and we deprecate `__experimentalFeatures`?

Reply:
Just double-checking, is this a question that needs an answer before this PR can land? If so, is it worth pinging for wider feedback?

My understanding is that by using `globalStylesLinksDataKey` it's effectively private(ish) for now 🤔

Reply:
Good question, sorry I missed this last night.

It's private and so I'm okay with it 😄 I was mainly wondering if there were any concerns about overloading. 

I guess we'll ask for forgiveness 😇

Reply:
What is this "links" data here? It's very unclear to me. I don't think we should be adding random block editor settings without a good reflexion on these. 

Aside: the useMemo is missing a dependency.

Reply:
The value comprises resolved, relative paths to files defined in theme.json. 

For example, if `/my-theme/image.png` is referenced as a background image in theme.json, the global styles REST API will return the full URI to the file in the `_links` so that the editor can correctly render the image. See https://github.com/WordPress/gutenberg/pull/61271

So, unclear maybe, but not random in my opinion.

The `Links` in `globalStylesLinksData` is an extension of `globalStylesData`, which, is a fairly opaque label in its own right.

If the value is unclear, then I think we should rename it. What about:

- `globalStylesResolvedURIs`
- `themeAssetPaths`
- ?

> Aside: the useMemo is missing a dependency.

Whoops, thanks for spotting that! 🙇🏻 

Update here:

- https://github.com/WordPress/gutenberg/pull/63823

Reply:
> I suppose it's safer than blindly passing on everything from mergedGlobalStyles, but wondering if they should be combined into a { styles, settings, _links } object and we deprecate __experimentalFeatures?

There was also this question above, as I was concerned about adding another property.

Reply:
> The Links in globalStylesLinksData is an extension of globalStylesData, which, is a fairly opaque label in its own right.

I agree that `globalStylesData` is an opaque label but it has a clear purpose: We're introducing support for a "theme" to block-editor package basically. 

> For example, if /my-theme/image.png is referenced as a background image in theme.json, the global styles REST API will return the full URI to the file in the _links object so that the editor can correctly render the image. See https://github.com/WordPress/gutenberg/pull/61271

Why aren't we instead replacing the urls in the endpoint in `globalStylesData` directly instead of having a separate `_links` property? What do we do for fonts and things like that?

Reply:
Thanks for questioning this. I spent a lot of time to-ing and fro-ing, so I admit I could be blinkered on the subject.

> Why aren't we instead replacing the urls in the endpoint in globalStylesData directly instead of having a separate _links property? 

That's the first approach I took, but ran into some snags.

The main reason I [rejected](https://github.com/WordPress/gutenberg/pull/61271#issuecomment-2099776380) that approach was because:

- when switching to a theme style variations, its contents are saved in the user global CPT in the database
- should a user want to export the theme, the relative paths should be preserved to enable theme portability

Resolving the image URIs on the backend once means there's no switching between the two: between relative and absolute, which would be especially tricky if the theme is a child theme referring to assets in some parent theme.

> What do we do for fonts and things like that?

I did look into fonts briefly, and we're using the [same syntax](https://github.com/WordPress/gutenberg/blob/b3c120a7e157dfaadf36d6911bbf5a706351bfd3/lib/compat/wordpress-6.4/fonts/font-face/class-wp-font-face-resolver.php#L148-L148) to represent.

What other assets work this way aside from fonts?

Also, I think fonts work differently in the way they're loaded. They're printed to the page once so their URLs can be resolved when printing out the [style tags](https://github.com/WordPress/gutenberg/blob/b3c120a7e157dfaadf36d6911bbf5a706351bfd3/lib/compat/wordpress-6.4/fonts/font-face/class-wp-font-face.php#L233-L233). 

Have the resolved background image URIs available in the frontend means we can apply them immediately without affecting the integrity of theme.json.

See for example how applying a variation that has a theme font needs a page reload to work:

https://github.com/user-attachments/assets/d9cce0d5-a9ed-4b71-be4b-3aedcb8cc7cd

cc @creativecoder for fact check

I'm happy to create an issue and investigate, if there's a specific issue to address here.

Reply:
> resolve the images properly before setting globalStylesData even if it's done in the frontend.

So long as `globalStylesData` isn't merged into the current user style object, and then saved to the database, it should be fine to find the images and replace them with the resolved path in `_links`. 

I'll check, but have a feeling I tried this and rolled it because it _was_ merged to the user style object. 

Also, I'd have to go back and double check the performance hit since the styles are rendered whenever a global styles is updated.

Reply:
> So long as globalStylesData isn't merged into the current user style object, and then saved to the database, it should be fine to find the images and replace them with the resolved path in _links.

Why is this such a big deal? I'm failing to understand, if you change a relative path with an absolute path, you're just making things clearer no?

Reply:
> Why is this such a big deal? I'm trying to understand, if you change a relative path with an absolute path, you're just making things clearer no?

It's only important if we care about preserving theme portability.

If an absolute path is saved to the database, then that absolute path will feature in the theme.json when a user exports their theme from the editor.

With the background image no longer referring to the exported theme's internal assets, it makes the exported theme very fragile and less portable. It places the onus on the user to dig into the JSON and fix the path.

I'm not dying on this hill by the way 😄 and am trying to think of alternatives, but I think they should satisfy the above requirement.

I also don't want to lose sight of the perceived area of concern — that editor settings has a new, private property that's related to the theme styles — and balancing the stakes here.
___ 

Sorry for the wall of text, but here's what I understand just to clarify:

1. Theme.json may refer to theme assets (images) using relative paths.
2.  The relative theme path could refer to a local asset, or a parent theme asset.
3. In order to build CSS in the editor (clientside) and on the frontend (PHP), we need to resolve the relative theme paths into absolute URLs.
4. The preference is to resolve the URLs on the backend using `get_theme_file_uri` as it will not only look at parent theme assets, but apply any filters.
5. Resolved URIs needs to be passed to the client (via REST) in order to build the correct CSS in the editor. 
6. To preserve theme.json style values, resolved URIs are embedded in `_links` in the response object.
7. In the client, the relative path -> absolute path substitution takes place when the CSS is built.
8. If a user applies a theme style variation that contains a relative path, that relative path is saved to the database, and follows the same process starting from point 1.

Now, with the way we've structured `_links` to include a theme.json path to the value, my guess is that it might be possible to replace the "urls in the endpoint in globalStylesData directly" as you mention, and then swap them back out before we commit any user styles to the global styles CPT in the database. To do this, we'd still need access to the `_links` object.

Reply:
I think you're right that the `_links` property make sense in the global styles endpoint. It gives the possibility to edit the styles while retaining the local aspect and add other local assets... This is something that can work conceptually.

I'm just thinking from the `block-editor` package's perspective, it might not make sense and I wonder if this discussion ties to the other discussion where I suggested as separate provider/package for Global Styles and whether that could simplify things here.

Obviously for now, this is private API, so it's "okey" but I would love if we can also follow-up with the right abstractions / right packages because we want third-parties to benefit from this as well.

---
Aside, I'm curious what's the shape of the `_links` object now?

Reply:
Great discussion, thanks for helping me think through it again. 

The `_links` object follows the documented standard for embedded URI resources:

https://developer.wordpress.org/rest-api/using-the-rest-api/linking-and-embedding/

```json
"_links": {
    "wp:theme-file-uris": [
        {
            "name": "file:/img/budgie3.jpg",
            "href": "http://ilovebudiges.com/wp-content/themes/emptytheme/img/budgie3.jpg",
            "target": "styles.background.backgroundImage.url",
            "type": "image/jpeg"
        }
    ]
}
```

"href", "type" and "name" are [HAL](https://en.wikipedia.org/wiki/Hypertext_Application_Language) reserved attributes. 

"target", a valid link attribute, represents the path in the theme.json tree to the value that needs to be replaced.

> follow-up with the right abstractions / right packages because we want third-parties to benefit from this as wel

The discussion over at https://github.com/WordPress/gutenberg/issues/63796 is sounding very productive. If I get down time during the next release cycle, I'd love to help out in any way.